### PR TITLE
Repairing incomplete consent records.

### PR DESCRIPTION
**Pull request by alxdarksage**
After authentication, look at the user's signatures. If one exists for a valid and active consent, but there's no consent record in DynamoDB, create it. We've seen a small (~5?) number of people from the early days of production where this has been the case.

### CODE REVIEW

File: app/org/sagebionetworks/bridge/services/AuthenticationService.java [hunk 1]
```diff
@@ -238,6 +254,40 @@ public void resetPassword(PasswordReset passwordReset) throws BridgeServiceExcep
         accountDao.resetPassword(passwordReset);
     }
     
+    private void repairConsents(Study study, Account account, UserSession session, ClientInfo clientInfo){
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/AuthenticationService.java:266]: Minor: Add some clarifying comments on why we need to do this, so future us don't look back on this code and get confused.

Some explanation about migrating from UserConsent2 to UserConsent3 and that wasn't completed successfully and how that left a bunch of accounts that have consent sigs in Stormpath but not in DDB, and that account.getActiveConsentSignature() gets consent sigs from Stormpath and how user.getConsentStatuses() gets consent sigs from DDB.

File: app/org/sagebionetworks/bridge/services/AuthenticationService.java [hunk 2]
```diff
@@ -238,6 +254,40 @@ public void resetPassword(PasswordReset passwordReset) throws BridgeServiceExcep
...
+        boolean repaired = false;
+        
+        Map<SubpopulationGuid,ConsentStatus> statuses = session.getUser().getConsentStatuses();
+        for (Map.Entry<SubpopulationGuid,ConsentStatus> entry : statuses.entrySet()) {
+            ConsentSignature activeSignature = account.getActiveConsentSignature(entry.getKey());
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/AuthenticationService.java:271]: Verify if account.getActiveConsentSignature() works for Stormpath accounts that have the old consent_signature key but not the new consent_signatureS key.

File: app/org/sagebionetworks/bridge/services/AuthenticationService.java [hunk 3]
```diff
@@ -238,6 +254,40 @@ public void resetPassword(PasswordReset passwordReset) throws BridgeServiceExcep
...
+            if (activeSignature != null && !status.isConsented()) {
+                repairConsent(session, entry.getKey(), activeSignature);
+                repaired = true;
+            }
+        }
+        
+        // These are incorrect since they are based on looking up DDB records, so re-create them.
+        if (repaired) {
+            updateConsentStatuses(study, session.getUser(), clientInfo);
+        }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/AuthenticationService.java:283]: Add a comment or rename the method to make it clear that this is updating the consent statuses in the session, which is a cache of the consents in DDB.

File: app/org/sagebionetworks/bridge/services/AuthenticationService.java [hunk 4]
```diff
@@ -238,6 +254,40 @@ public void resetPassword(PasswordReset passwordReset) throws BridgeServiceExcep
...
+    }
+    
+    /**
+     * If the signature exists but consentStatus says the user is not consented, then the record is missing in
+     * DDB. We need to create it.
+     */
+    private void repairConsent(UserSession session, SubpopulationGuid subpopGuid, ConsentSignature activeSignature) {
+        logger.error("Signature found without a matching user consent record. Adding consent for " + session.getUser().getId());
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/AuthenticationService.java]: Is this an error, a warning, or info? Error suggests this indicates that something is wrong with the server and devs should look into it. Warning suggests that a few of these is not a problem, but a lot of them indicates a systemic problem. Info suggests it's worth looking at it, but it shouldn't raise any alarms.

File: app/org/sagebionetworks/bridge/services/AuthenticationService.java [hunk 5]
```diff
@@ -238,6 +254,40 @@ public void resetPassword(PasswordReset passwordReset) throws BridgeServiceExcep
...
+        long signedOn = activeSignature.getSignedOn();
+        if (signedOn == 0L) {
+            signedOn = DateTimeUtils.currentTimeMillis(); // this is so old we did not record a signing date...
+        }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/AuthenticationService.java:295]: Is it possible to test this specific case? Evidence suggests that all affected users do not have a signedOn in their Stormpath consent sig. Ergo, this is the most important case to test.

File: app/org/sagebionetworks/bridge/services/AuthenticationService.java (hunk 3 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/services/AuthenticationService.java:283]: renamed to updateInMemoryConsentStatuses

File: app/org/sagebionetworks/bridge/services/AuthenticationService.java (hunk 1 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/services/AuthenticationService.java:266]: Added

File: app/org/sagebionetworks/bridge/services/AuthenticationService.java (hunk 4 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/services/AuthenticationService.java]: I'd like to know if this happens. I don't think we need to do anything about it, we'd just do what we're automating here. But I would like to look and see if users are still being created in this state, or if this is only a historical problem. That is, some users have had a DynamoUserConsent2 record, but some have not.

File: app/org/sagebionetworks/bridge/services/AuthenticationService.java (hunk 5 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/services/AuthenticationService.java:295]: In the test for this, I zeroed out the signature, and deleted the history putting the signature in the singular field slot, as was done before implementing consent withdrawal. The repair method still repairs the account correctly.

**Comment by DwayneJengSage** [comment]
The log message in AuthenticationService.repairConsent() should be an info, not an error. Errors and warnings will trigger alarms, and it sounds like we don't need or want to trigger alarms when that happens.

File: app/org/sagebionetworks/bridge/services/AuthenticationService.java (hunk 2 above)
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/services/AuthenticationService.java:271]: Added to test

**Pull request merged**